Centralize a bit, track render pass dependencies better

unknownbrackets opened the pull request:
This takes most of the original #12913, without actually doing any invalidation on GLES.

While this changes INVALIDATE, nothing actually triggers that right now.  Also it was buggy and wrong before (used the wrong bits for stencil, etc.)  No reason to keep it wrong.

This also keeps the dependency tracking added to GLES, but doesn't use it for anything yet.

-[Unknown]

unknownbrackets committed:
GPU: Centralize framebuffer download.

And try not to use a potentially-null nvfb.  Primarily this can happen on
Direct3D backends.

unknownbrackets committed:
GLES: Add dependency tracking for render passes.

Let's try to invalidate when it's possible.  We move the invalidate to the
end of the render when detected.

unknownbrackets committed:
Vulkan: Generalize dependency tracking.

Might as well track on all steps.  We also can know if the dst was fully
written to here.

unknownbrackets committed:
GLES: Fix invalidation of backbuffer.

And of the stencil buffer.

hrydgard commented:
Thanks for splitting it out!

I'll just check this with God of War and wait for CI... which seems a little stuck or overloaded right now.

hrydgard commented:
This did affect how many render passes GoW is able to merge, still merging a lot but less than before (probably better for correctness, but ...).

I'm gonna try to figure out what changed before merging.

hrydgard commented on ext/native/thin3d/VulkanRenderManager.cpp:
```diff
@@ -490,6 +490,13 @@ void VulkanRenderManager::BindFramebufferAsRenderTarget(VKRFramebuffer *fb, VKRR
 	step->render.finalColorLayout = !fb ? VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL : VK_IMAGE_LAYOUT_UNDEFINED;
 	steps_.push_back(step);
 
+	if (fb) {
+		// If there's a KEEP, we naturally read from the framebuffer.
+		if (color == VKRRenderPassAction::KEEP || depth == VKRRenderPassAction::KEEP || stencil == VKRRenderPassAction::KEEP) {
+			step->dependencies.insert(fb);
+		}
+	}
```
Commenting out this part restores the render pass merging in GoW to what it was before (graphics appear to work fine either way).

Not quite sure how the new general check would stop this but not the old ones...

hrydgard commented:
[God of War rp merge test UCES01401_0001.zip](https://github.com/hrydgard/ppsspp/files/4640933/God.of.War.rp.merge.test.UCES01401_0001.zip)

A dump of the scene I'm testing... I'll investigate more later, done enough hacking today...

unknownbrackets commented:
Checking.  Also, realized we actually can't merge past the dests even if clobbered without renaming, bah.  So this probably should not change merging at all.  Still, I think tracking read dependencies generally will simplify other optimizations.

We may also be able to move others up (the ones that are dependent.)

-[Unknown]

unknownbrackets commented:
It's because `touchedFramebuffers` contains those framebuffers and they are using KEEP.

Merging step 0, stopped at step 3 because it continues FB touched by step 1
Merging step 1, stopped at step 6 because it continues FB touched by step 4
Merging step 8, stopped at step 11 because it continues FB touched by step 9
Merging step 9, stopped at step 12 because it continues FB touched by step 10
Merging step 10, stopped at step 15 because it continues FB touched by step 13
Merging step 13, stopped at step 18 because it continues FB touched by step 16
Merging step 16, stopped at step 21 because it continues FB touched by step 19
Merging step 19, stopped at step 24 because it continues FB touched by step 22
Merging step 22, stopped at step 27 because it continues FB touched by step 25

With that said, I'm not sure we need to stop.  Thinking this through...

-[Unknown]

unknownbrackets committed:
Vulkan: Make merge pass a bit easier to read.

unknownbrackets committed:
Vulkan: Stop merge at clear/touched read only.

Before we would merge on a future render after a clear, which can't be
good.  But we can safely ignore a draw to another FB that reads from a
touched FB.

unknownbrackets commented:
Okay, that GE capture now merges just as well as before.

This may merge more or less in other games/scenes, though, and should be more correct.  Specifically:

1. In a scenario (all same target FB) of KEEP, CLEAR, KEEP... we would merge the two KEEPs (same probablem with DONT_CARE.)  This is no longer allowed, the CLEAR is a barrier.

2. We no longer stop when a pass targeting another FB depends on a touched FB, as long as it doesn't depend on the merge target.

I think there's more merging opportunity available, though.  Specifically thinking of:

1. Draw to A (CLEAR, no deps) // Initial scene
2. Draw to B (CLEAR, no deps) // Sign, billboard, shadow, whatever
3. Draw to A (KEEP, dep on A + B) // More of scene (with whatever)
4. Draw to B (KEEP, dep on B) // Other sign
5. Draw to A (KEEP, dep on A + B) // More of scene (with sign)

In this scenario, we can't merge 3 into 1, but we could merge 1 into 3.  We already stop when we hit a dep on self.  Not sure if I'm missing any other case...

-[Unknown]

hrydgard commented:
Awesome!

We don't really need to absolutely maximize merging, the important thing is the pathological cases like the one demonstrated, that's where the big benefit is :)  Still of course, good to take opportunities that are there. Aliasing/renaming will open up some more as discussed, with more work and more bug potential :)

unknownbrackets commented:
Might be NVIDIA does this sort of merging automatically somehow, though maybe it's just so fast it doesn't matter.  But yeah, of course.  I assume tiled renderers would get a lot out of merging to begin with.

-[Unknown]
